pkg/controllers/util/federatedclient/podinformer_test.go

```diff
@@ -0,0 +1,346 @@
+/*
+Copyright 2023 The KubeAdmiral Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package federatedclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"golang.org/x/sync/semaphore"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/rand"
+	kubeinformer "k8s.io/client-go/informers"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/fake"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	fakecorev1 "k8s.io/client-go/kubernetes/typed/core/v1/fake"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/kubewharf/kubeadmiral/pkg/controllers/util"
+)
+
+var errLimitExceeded = errors.New("limit exceeded")
+
+type admiralFakePods struct {
+	corev1client.PodInterface
+	concurrentLists *atomic.Int64
+}
+
+func (c *admiralFakePods) List(ctx context.Context, opts metav1.ListOptions) (result *corev1.PodList, err error) {
+	cur := c.concurrentLists.Add(-1)
+	defer c.concurrentLists.Add(1)
+	if cur < 0 {
+		panic(errLimitExceeded)
+	}
+	return c.PodInterface.List(ctx, opts)
+}
+
+type admiralFakeCoreV1 struct {
+	fakecorev1.FakeCoreV1
+	concurrentLists *atomic.Int64
+}
+
+func (c *admiralFakeCoreV1) Pods(namespace string) corev1client.PodInterface {
+	return &admiralFakePods{c.FakeCoreV1.Pods(namespace), c.concurrentLists}
+}
+
+var _ kubeclient.Interface = &admiralFakeClientset{}
+
+type admiralFakeClientset struct {
+	*fake.Clientset
+	concurrentLists *atomic.Int64
+}
+
+func (c *admiralFakeClientset) CoreV1() corev1client.CoreV1Interface {
```

Comment:
```suggestion
func (c *fakeClientsetWithConcurrencyLimit) CoreV1() typedcorev1.CoreV1Interface {
```
`admiralFakeClientset` could mean the "fake clientset for the admiral api".